### ISSUE TITLE
[lncli] rpc error: code = Unknown desc = connection attempt to (node)@ip:9735 is pending

### ISSUE THREAD

**Issue by robtex** [lightningnetwork/lnd#941]
[lncli] rpc error: code = Unknown desc = connection attempt to (node)@ip:9735 is pending

getting this when trying to connect to some nodes
it seems the lnd is already trying to connect to the remote node (but fails)
after checking with tcpdump i can see it does, but it uses the wrong destination port.
probably this is because lnd tries to connect to the *source*port of the remote node from a previous *inbound* connection.
no idea how to workaround this in a good way. i managed to do it in a bad way by using iptables to correct the outgoing connections to connect to port 9735 instead of random port.

correct fix should be imho to
1) only try to connect to remote nodes that were outbound, to make sure it is a correct port
2) make sure these reconnects give up and gets removed after a while

**Comment by vegardengen** [comment]
Fixing this so that you can override this, and that a lncli connect always overrides the pending connections, was pretty easy. I can now with my small fix do lncli connect and have it override pending connections.

Downsides: If the lncli connect fails, there's no longer any automated reconnection effort. This might not be so bad if it was connecting to the wrong port, but it was connecting to the right port and you did an lncli connect, you'd have it override automatic reconnection...

Maybe check if you are connecting to same host/port, and in that case, not cancel the pending connection and not try connecting (i.e. revert to previous behavior if you are trying to connect to same host/port as a pending reconnect)?

**Comment by robtex** [comment]
fyi the reconnection database seems to have duplicate entries of the same destination. 
my lnd node is currectly sending out 20 SYN per second or so in total.
to one victim my lnd is sending packets to 300 different ports, 4 per second, although it already has a proper connection on :9735 
i think prioritizing this bug would be good since it will solve many other peculiar bugs which are sideeffects from this root cause.

**Comment by halseth** [comment]
Related: #935

**Comment by robtex** [comment]
i'm currently trying out this patch in peerTerminationWatcher to limit reconnects to wrong port 
```
        // Add this. patch by vegard/rompert
        if !p.inbound { // node that inbound logic is still reversed
                srvrLog.Debugf("Not Scheduling connection re-establishment to "+
                        "inbound peer %v", p)
                return                                                                                                                                                               }
```
it is far from a complete fix though

**Comment by robtex** [comment]
without that patch it means that when lnd receives a message it doesn't like (address type 206/27) and hangs up, lnd disconnected and instantly tries to reconnect to a random port
the other node then tries to reestablish a new connection
rince and repeat and you get a lot of connections. i've seen 20 per second in total and 4 per second to a single ip number

**Comment by robtex** [comment]
also this patch helps you help lnd to connect to the correct port when it tries the wrong one and blocks you from trying due to "pending" (which effectively was deadlocking two lnd nodes from communicating with eachother)

https://github.com/vegardengen/lnd/commit/a138b9c1ae6ada003679598579d8ca69c6e4059b

```
@@ -1710,11 +1710,14 @@ func (s *server) ConnectToPeer(addr *lnwire.NetAddress, perm bool) error {
 	// Peer was not found, continue to pursue connection with peer.
 
 	// If there's already a pending connection request for this pubkey,
-	// then we ignore this request to ensure we don't create a redundant
-	// connection.
-	if _, ok := s.persistentConnReqs[targetPub]; ok {
-		s.mu.Unlock()
-		return fmt.Errorf("connection attempt to %v is pending", addr)
+	// remove the pending connection requests.
+	if connReqs, ok := s.persistentConnReqs[targetPub]; ok {
+		for _, pConnReq := range connReqs {
+			if pConnReq != nil {
+				s.connMgr.Remove(pConnReq.ID())
+			}
+		}
+		delete(s.persistentConnReqs, targetPub)
 	}
 
 	// If there's not already a pending or active connection to this node,
```

**Comment by Roasbeef** [comment]
Fixed by #982.
